Firestore/core/test/firebase/firestore/local/leveldb_mutation_queue_test.cc

```diff
@@ -0,0 +1,188 @@
+const char* kDummy = "1";
+
+std::unique_ptr<Persistence> PersistenceFactory() {
+  return LevelDbPersistenceForTesting();
+}
+
+}  // namespace
+
+INSTANTIATE_TEST_CASE_P(LevelDbMutationQueueTest,
+                        MutationQueueTest,
+                        testing::Values(PersistenceFactory));
+
+class LevelDbMutationQueueTest : public MutationQueueTestBase {
+ public:
+  LevelDbMutationQueueTest()
+      : MutationQueueTestBase(PersistenceFactory()),
+        db_(static_cast<LevelDbPersistence*>(persistence_.get())->ptr()) {
+  }
+
+ protected:
+  void SetDummyValueForKey(const std::string& key);
+
+  DB* db_;
+};
+
+/**
+ * Creates a key that's structurally the same as LevelDbMutationKey except it
+ * allows for nonstandard table names.
+ */
+std::string MutationLikeKey(absl::string_view table,
+                            absl::string_view user_id,
+                            BatchId batch_id) {
+  std::string key;
+  OrderedCode::WriteSignedNumIncreasing(&key, 5);  // TableName
+  OrderedCode::WriteString(&key, table);
+  OrderedCode::WriteSignedNumIncreasing(&key, 13);  // UserId
+  OrderedCode::WriteString(&key, user_id);
+  OrderedCode::WriteSignedNumIncreasing(&key, 10);  // BatchId
+  OrderedCode::WriteSignedNumIncreasing(&key, batch_id);
+  OrderedCode::WriteSignedNumIncreasing(&key, 0);  // Terminator
+  return key;
+}
+
+TEST_F(LevelDbMutationQueueTest, LoadNextBatchID_zeroWhenTotallyEmpty) {
+  // Initial seek is invalid
+  ASSERT_EQ(LoadNextBatchIdFromDb(db_), 1);
+}
+
+TEST_F(LevelDbMutationQueueTest, LoadNextBatchID_zeroWhenNoMutations) {
+  // Initial seek finds no mutations
+  SetDummyValueForKey(MutationLikeKey("mutationr", "foo", 20));
+  SetDummyValueForKey(MutationLikeKey("mutationsa", "foo", 10));
+  ASSERT_EQ(LoadNextBatchIdFromDb(db_), 1);
+}
+
+TEST_F(LevelDbMutationQueueTest, LoadNextBatchID_findsSingleRow) {
+  // Seeks off the end of the table altogether
+  SetDummyValueForKey(LevelDbMutationKey::Key("foo", 6));
+
+  ASSERT_EQ(LoadNextBatchIdFromDb(db_), 7);
+}
+
+TEST_F(LevelDbMutationQueueTest,
+       LoadNextBatchID_findsSingleRowAmongNonMutations) {
+  // Seeks into table following mutations.
+  SetDummyValueForKey(LevelDbMutationKey::Key("foo", 6));
+  SetDummyValueForKey(MutationLikeKey("mutationsa", "foo", 10));
+
+  ASSERT_EQ(LoadNextBatchIdFromDb(db_), 7);
+}
+
+TEST_F(LevelDbMutationQueueTest, LoadNextBatchID_findsMaxAcrossUsers) {
+  SetDummyValueForKey(LevelDbMutationKey::Key("fo", 5));
+  SetDummyValueForKey(LevelDbMutationKey::Key("food", 3));
+
+  SetDummyValueForKey(LevelDbMutationKey::Key("foo", 6));
+  SetDummyValueForKey(LevelDbMutationKey::Key("foo", 2));
+  SetDummyValueForKey(LevelDbMutationKey::Key("foo", 1));
+
+  ASSERT_EQ(LoadNextBatchIdFromDb(db_), 7);
+}
+
+TEST_F(LevelDbMutationQueueTest, LoadNextBatchID_onlyFindsMutations) {
+  // Write higher-valued batch_i_ds in nearby "tables"
+  std::vector<std::string> tables{"mutatio", "mutationsa", "bears", "zombies"};
```

Comment:
Mutations, zombies... Feels like Resident Evil. No idea what the bears are doing here.

Reply:
Stephen Colbert (on his satirical show, _The Colbert Report_) famously called bears "godless killing machines" and "the greatest threat to America".